src/query/functions/aggregation/take.go

```diff
@@ -0,0 +1,176 @@
+		return baseOp{}, fmt.Errorf("operator not supported: %s", opType)
+	}
+
+	heap := utils.NewFloatHeap(takeTop, int(params.Parameter))
+	fn := func(values []float64, buckets [][]int) []float64 {
+		return takeFn(heap, values, buckets)
+	}
+
+	return newTakeOp(params, opType, fn), nil
+}
+
+// takeOp stores required properties for take ops
+type takeOp struct {
+	params   NodeParams
+	opType   string
+	takeFunc takeFunc
+}
+
+// OpType for the operator
+func (o takeOp) OpType() string {
+	return o.opType
+}
+
+// String representation
+func (o takeOp) String() string {
+	return fmt.Sprintf("type: %s", o.OpType())
+}
+
+// Node creates an execution node
+func (o takeOp) Node(
+	controller *transform.Controller,
+	_ transform.Options,
+) transform.OpNode {
+	return &takeNode{
+		op:         o,
+		controller: controller,
+	}
+}
+
+func newTakeOp(params NodeParams, opType string, takeFunc takeFunc) takeOp {
+	return takeOp{
+		params:   params,
+		opType:   opType,
+		takeFunc: takeFunc,
+	}
+}
+
+// takeNode is different from base node as it only uses grouping to determine
+// groups from which to take values from, and does not necessarily compress the
+// series set as regular aggregation functions do
+type takeNode struct {
+	op         takeOp
+	controller *transform.Controller
+}
+
+// Process the block
+func (n *takeNode) Process(ID parser.NodeID, b block.Block) error {
+	stepIter, err := b.StepIter()
+	if err != nil {
+		return err
+	}
+
+	params := n.op.params
+	meta := stepIter.Meta()
+	seriesMetas := utils.FlattenMetadata(meta, stepIter.SeriesMeta())
+	buckets, _ := utils.GroupSeries(
+		params.MatchingTags,
+		params.Without,
+		n.op.opType,
+		seriesMetas,
+	)
+
+	// retain original metadatas
+	builder, err := n.controller.BlockBuilder(meta, stepIter.SeriesMeta())
+	if err != nil {
+		return err
+	}
+
+	if err := builder.AddCols(stepIter.StepCount()); err != nil {
+		return err
+	}
+
+	for index := 0; stepIter.Next(); index++ {
+		step, err := stepIter.Current()
+		if err != nil {
+			return err
+		}
+
+		values := step.Values()
+		aggregatedValues := n.op.takeFunc(values, buckets)
+		builder.AppendValues(index, aggregatedValues)
+	}
+
+	nextBlock := builder.Build()
+	defer nextBlock.Close()
+	return n.controller.Process(nextBlock)
+}
+
+func takeFn(heap utils.FloatHeap, values []float64, buckets [][]int) []float64 {
+	cap := heap.Cap()
+	for _, bucket := range buckets {
+		// If this bucket's length is less than or equal to the heap's
+		// capacity do not need to clear any values from the input vector,
+		// as they are all included in the output.
+		if len(bucket) <= cap {
+			continue
+		}
+
+		// Add values from this bucket to heap, clearing them from input vector
+		// after they are in the heap.
+		for _, idx := range bucket {
+			val := values[idx]
+			if !math.IsNaN(val) {
+				heap.Push(values[idx], idx)
+			}
+
+			values[idx] = math.NaN()
+		}
+
+		// Re-add the val/index pairs from the heap to the input vector
```

Comment:
how do you match series metadata ? Seems like for every column, you can have a different ordering of values ?

Reply:
The value order does not change. The index of each value added to the heap is stored as well, and these are added to the output value list at the same index that they were taken from.

TakeK functions keep the metadata of the incoming series, so these will always be in order.

Reply:
I see. So for 100 series, if I take top 10, then for each time interval, the series for which datapoints are in top 10 will be set and other will be NaNs ?